pkg/cmd/server/kubernetes/node.go

```diff
@@ -126,17 +126,29 @@ func (c *NodeConfig) initializeVolumeDir(ce commandExecutor, path string) (strin
 
 // RunKubelet starts the Kubelet.
 func (c *NodeConfig) RunKubelet() {
-	// TODO: clean this up and make it more formal (service named 'dns'?). Use multiple ports.
-	clusterDNS := c.KubeletConfig.ClusterDNS
-	if clusterDNS == nil {
-		if service, err := c.Client.Endpoints(kapi.NamespaceDefault).Get("kubernetes"); err == nil {
-			if ip, ok := firstIP(service, 53); ok {
-				if err := cmdutil.WaitForSuccessfulDial(false, "tcp", fmt.Sprintf("%s:%d", ip, 53), 50*time.Millisecond, 0, 2); err == nil {
-					c.KubeletConfig.ClusterDNS = net.ParseIP(ip)
+	if c.KubeletConfig.ClusterDNS == nil {
+		// Set ClusterDNS to "kubernetes" master service ip if port 53/tcp can be successfully dialed
+		if service, err := c.Client.Services(kapi.NamespaceDefault).Get("kubernetes"); err == nil {
+			if includesServicePort(service.Spec.Ports, 53, "dns-tcp") {
+				serviceIP := service.Spec.ClusterIP
+				if err := cmdutil.WaitForSuccessfulDial(false, "tcp", fmt.Sprintf("%s:%d", serviceIP, 53), 50*time.Millisecond, 0, 2); err == nil {
```

Comment:
Just use it

On Oct 29, 2015, at 8:00 AM, Jordan Liggitt <notifications@github.com>
wrote:

In pkg/cmd/server/kubernetes/node.go
<https://github.com/openshift/origin/pull/5094#discussion_r43376225>:

> @@ -126,17 +126,29 @@ func (c *NodeConfig) initializeVolumeDir(ce commandExecutor, path string) (strin
>
>  // RunKubelet starts the Kubelet.
>  func (c *NodeConfig) RunKubelet() {
> -	// TODO: clean this up and make it more formal (service named 'dns'?). Use multiple ports.
> -	clusterDNS := c.KubeletConfig.ClusterDNS
> -	if clusterDNS == nil {
> -		if service, err := c.Client.Endpoints(kapi.NamespaceDefault).Get("kubernetes"); err == nil {
> -			if ip, ok := firstIP(service, 53); ok {
> -				if err := cmdutil.WaitForSuccessfulDial(false, "tcp", fmt.Sprintf("%s:%d", ip, 53), 50*time.Millisecond, 0, 2); err == nil {
> -					c.KubeletConfig.ClusterDNS = net.ParseIP(ip)
> +	if c.KubeletConfig.ClusterDNS == nil {
> +		// Set ClusterDNS to "kubernetes" master service ip if port 53/tcp can be successfully dialed
> +		if service, err := c.Client.Services(kapi.NamespaceDefault).Get("kubernetes"); err == nil {
> +			if includesServicePort(service.Spec.Ports, 53, "dns-tcp") {
> +				serviceIP := service.Spec.ClusterIP
> +				if err := cmdutil.WaitForSuccessfulDial(false, "tcp", fmt.Sprintf("%s:%d", serviceIP, 53), 50*time.Millisecond, 0, 2); err == nil {

@smarterclayton <https://github.com/smarterclayton>, if they specified a
port named "dns" on port 53, should we even ping it or just use it?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5094/files#r43376225>.